service/gcs/core/gcs/gcs.go

```diff
@@ -288,24 +302,28 @@ func (c *gcsCore) ExecProcess(id string, params prot.ProcessParameters, stdioSet
 	// apply to the old process no longer makes sense, so since the old
 	// process's pid has been reused, its cache entry can also be reused.  This
 	// applies to external processes as well.
-	c.processCache[pid] = processEntry
+	c.processCache[p.Pid()] = processEntry
 	c.processCacheMutex.Unlock()
-	containerEntry.AddProcess(pid)
-	return pid, nil
+	containerEntry.AddProcess(p.Pid())
+	return p.Pid(), nil
 }
 
 // SignalContainer sends the specified signal to the container's init process.
 func (c *gcsCore) SignalContainer(id string, signal oslayer.Signal) error {
 	c.containerCacheMutex.Lock()
 	defer c.containerCacheMutex.Unlock()
 
-	if _, ok := c.containerCache[id]; !ok {
+	containerEntry := c.getContainer(id)
+	if containerEntry == nil {
 		return errors.WithStack(gcserr.NewContainerDoesNotExistError(id))
 	}
 
-	if err := c.Rtime.KillContainer(id, signal); err != nil {
-		return err
+	if containerEntry.container != nil {
```

Comment:
That would be nice, agreed. I'm not sure that the runc package is the best place to make that distinction, though, because runc itself doesn't have the concept of a container without an initial process. We can think about it some more -- my intention with this change is not to change the underlying method flow, just to make it a little more object-oriented. We can change the flow later if we think it will make things cleaner.